core/src/main/java/eu/fasten/core/data/opal/MavenArtifactDownloader.java

```diff
@@ -0,0 +1,121 @@
+package eu.fasten.core.data.opal;
+
+import eu.fasten.core.data.opal.exceptions.MissingArtifactException;
+import eu.fasten.core.maven.utils.MavenUtilities;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.util.Arrays;
+import java.util.Optional;
+
+/**
+ * A set of methods for downloading POM and JAR files given Maven coordinates.
+ */
+public class MavenArtifactDownloader {
+    private static final Logger logger = LoggerFactory.getLogger(MavenArtifactDownloader.class);
+    private static final String[] packaging = {"jar", "war", "zip", "ear", "rar", "ejb", "par",
+            "aar", "car", "nar", "kar"};
+    private static final String[] defaultPackaging = {"zip", "aar", "tar.gz", "jar"};
+
+    /**
+     * Download a JAR file indicated by the provided Maven coordinate.
+     *
+     * @param mavenCoordinate A Maven coordinate in the for "groupId:artifactId:version"
+     * @return A temporary file on the filesystem
+     */
+    public File downloadArtifact(final MavenCoordinate mavenCoordinate, String artifactRepo)
+            throws MissingArtifactException {
+        var found = false;
+        Optional<File> jar = Optional.empty();
+        var repos = mavenCoordinate.getMavenRepos();
+        if (artifactRepo != null && !artifactRepo.isEmpty() && !artifactRepo.equals(MavenUtilities.MAVEN_CENTRAL_REPO)) {
+            repos.addFirst(artifactRepo);
+        }
+        for (int i = 0; i < repos.size(); i++) {
+
+            long startTime = System.nanoTime();
+
+            try {
+                if (Arrays.asList(packaging).contains(mavenCoordinate.getPackaging())) {
+                    found = true;
+                    jar = httpGetFile(mavenCoordinate
+                            .toProductUrl(repos.get(i), mavenCoordinate.getPackaging()));
+                }
+            } catch (MissingArtifactException e) {
```

Comment:
As we are cycling through multiple repositories, the log message should boil down to a single info entry that explains that artifacts was not found in repo XY and that we retry the next one... if it is not found in any repo, we can throw a `MissingArtifactException` at the end.

Related to that, I noticed that several repos are being tried before Maven Central, but as it is the most common, I would argue that Maven Central should be the first hit in the list... I am not sure though where in the code the repos are hard-coded though.

Reply:
Source repos are specified in the input records of OPAL, i.e., the `artifactRepository` field. I think it tries the Maven Central repo first most of the time as most input records have Maven Central defined as the source repo.